### ISSUE TITLE
Add an option to post eval result and stdout/stderr result as comments

### ISSUE THREAD

**Issue by haya14busa** [nwtgck/actions-comment-run#112]
How about addding an option to post eval result and stdout/stderr result as comments?

For example, how about adding `-comment` flag? It should be useful especially for non-js code because it's not easy to post comments to GitHub.

@github-actions run -comment

```sh
#!/bin/sh
go get github.com/erning/gorun
```

```go
#!/home/runner/go/bin/gorun

package main

import "fmt"

func main() {
	fmt.Println("hello, world")
}
```

**Comment by nwtgck** [comment]
@haya14busa Hi!

Thank you very much for your wonderful idea! I also think flags are very useful. I'll design the flags. I prefer to use `--comment` than using single `-`.

Maybe it is better to use as follows because users can select one of them or both of them.
- `--stdout=comment`
- `--stderr=comment`
The disadvantage of `--stdout=...` and `--stderr=...` is typing a lot.

I'll consider of the design the flags. When you or other visitors have ideas, let me know.

**Comment by haya14busa** [comment]
Cool. Maybe we can add short flag (e.g. `-c`) as well in addition to the long flags.

**Comment by nwtgck** [comment]
That's right! I'd like introduce long and short flags.

**Comment by ibnesayeed** [comment]
A rather simpler approach would be to make `postcomment` CLI command available in this environment that posts contents of a file or stream as comment. This will be analogous to `postComment` JS function, but will be available as an executable process instead. This way, people can decide what they want to comment, irrespective of the language they use. Having flags to post `STDOUT`/`STDERR` means it will be an all or nothing situation. What if we only want to post certain outputs, not everything?

Here is what proposed CLI script should be able to do in a more flexible manner:

@github-actions run

```sh
#!/bin/sh
echo "Show this in logs, not in comments"
echo "Post this as a comment" > /tmp/cmnt-1.txt
echo "Appended comment line" >> /tmp/cmnt-1.txt
echo "Show both in logs and comments" | tee -a /tmp/cmnt-1.txt
echo "Another comment text" > /tmp/cmnt-2.txt
postcomment /tmp/cmnt-*.txt
echo "Post yet another comment separately" | postcomment
```

**Comment by nwtgck** [comment]
Thanks.

Do you mean some special files are posted as GitHub issue comment or define another special issue comment like`@github-actions comment`?

**Comment by ibnesayeed** [comment]
@nwtgck I meant, the way `cat` is a command that concatenates contents of one or more files or input stream and `sort` is a command that sorts lines of one or more files or input stream to output it on the `STDOUT`. The same way, we can introduce an executable script named `postcomment` (simply a wrapper around the `postComment` JS function) and add it in the path.

**Comment by nwtgck** [comment]
@ibnesayeed Thanks. I wouldn't like to install an additional command as possible for maintainability and it's destructive.

**Comment by ibnesayeed** [comment]
Well, we could use [GitHub's official CLI](https://cli.github.com/manual/) that is already available in the runner environment, but I am not sure if it allows commenting on issues.

**Comment by nwtgck** [comment]
Very good information!

**Comment by nwtgck** [comment]
I found the issue https://github.com/cli/cli/issues/517.

**Comment by nwtgck** [comment]
CLI is also great, but it is easy to use only from shell. The action runs any language with shebang, so I'd like to find another way to post an issue comment from any language.

**Comment by ibnesayeed** [comment]
While we can wait for the official CLI to add support for comments, I do not necessarily agree with adding a command being destructive as long as used name is not too generic to have any conflicts with existing environment. We are not replacing anything, we are simply adding yet another script and adding its location in the `PATH`. Since we can reuse the existing JS function to write the script, cross platform compatibility would not be that difficult. Besides, in the example YAML file we are proposing to use Ubuntu as the runner anyway. Finally, using. `comment` in place of `run` does not really solve the issue of picking and choosing what to post and what not to post, in addition to that it sounds like we are not running anything, just commenting.

**Comment by nwtgck** [comment]
> Finally, using. comment in place of run does not really solve the issue of picking and choosing what to post and what not to post, in addition to that it sounds like we are not running anything, just commenting.

Yes. In that situation, we should write program for it on GitHub issue comment. I'd like to keep this action simple.

**Comment by ibnesayeed** [comment]
> CLI is also great, but it is easy to use only from shell. The action runs any language with shebang, so I'd like to find another way to post an issue comment from any language.

Every language has a way to execute CLI commands, so using CLI covers them all. Other alternative would be to implement helper function for every environment, which would be a bigger hammer.

Thinking out loud, can we assign a special file in a very specific location where a script in any language can place any content, that will then be read and posted as a comment towards the end of the action as the cleanup process? Alternatively, can we leverage a well-defined [output parameter](https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-output-parameter) (say, `comment_text`) who's value, if not empty, can be posted towards the end of the action?

**Comment by nwtgck** [comment]
@ibnesayeed 

> Every language has a way to execute CLI commands, so using CLI covers them all.

Yes. Almost languages has features to run external commands. But I'd like to have easier way to post comments. If you want use the CLI, you can install as follows.

```yaml
    - run: <INSTALL posting comment CLI from somewhere>
    - uses: nwtgck/actions-comment-run@v1.1
      with:
        github-token: ${{ secrets.GITHUB_TOKEN }}
        allowed-associations: '["OWNER"]'
```

**Comment by ibnesayeed** [comment]
I know I can create and install such a CLI as a separate action, but its usefulness beyond this action will be limited. That said, it was a suggestion to solve the problem at hand in a rather flexible manner, if you do not find it appealing then that's alright.

**Comment by nwtgck** [comment]
I made a simple shell script to post comment. This might be helpful to create a simple CLI.

## demo

<img width="898" alt="image" src="https://user-images.githubusercontent.com/10933561/83932703-82123b80-a7df-11ea-9bee-42d2b9ec307b.png">


## comment

````md
@github-actions run

```bash
#! /bin/bash

GITHUB_TOKEN=$(node -p "process.env['INPUT_GITHUB-TOKEN']")
REPO_FULL_NAME=$(cat $GITHUB_EVENT_PATH | jq -r .repository.full_name)
ISSUE_NUMBER=$(cat $GITHUB_EVENT_PATH | jq .issue.number)
URL="https://api.github.com/repos/${REPO_FULL_NAME}/issues/${ISSUE_NUMBER}/comments"

echo '{"body": "hello world"}' | curl -u github-actions:${GITHUB_TOKEN} -T - -X POST $URL
```
````

**Comment by nwtgck** [comment]
@ibnesayeed 

Finally, I made `postcomand` in this action.

<img width="904" alt="image" src="https://user-images.githubusercontent.com/10933561/83933455-9bb58200-a7e3-11ea-9c4f-2c67b7008e5f.png">

<img width="917" alt="image" src="https://user-images.githubusercontent.com/10933561/83933461-a3752680-a7e3-11ea-85e5-90403d3f6f28.png">


## Raw comment

````md
@github-actions run

## Install `postcomment`

```bash
#! /bin/bash

sudo cat << 'EOF' | sudo tee /usr/local/bin/postcomment
#! /bin/bash

GITHUB_TOKEN=$(node -p "process.env['INPUT_GITHUB-TOKEN']")
REPO_FULL_NAME=$(cat $GITHUB_EVENT_PATH | jq -r .repository.full_name)
ISSUE_NUMBER=$(cat $GITHUB_EVENT_PATH | jq .issue.number)
URL="https://api.github.com/repos/${REPO_FULL_NAME}/issues/${ISSUE_NUMBER}/comments"

for FILE in "$@"
do
  jq -n --arg body "$(cat $FILE)" '{body: $body}' | curl -u github-actions:${GITHUB_TOKEN} -T - -X POST $URL
done

EOF

sudo chmod +x /usr/local/bin/postcomment
```

## Main

```bash
#!/bin/bash

echo "Show this in logs, not in comments"
echo "Post this as a comment" > /tmp/cmnt-1.txt
echo "Appended comment line" >> /tmp/cmnt-1.txt
echo "Show both in logs and comments" | tee -a /tmp/cmnt-1.txt
echo "Another comment text" > /tmp/cmnt-2.txt
postcomment /tmp/cmnt-*.txt
```
````


You can use saved-replies functionality to easy to use this snippet like the following.

![](https://github.com/nwtgck/actions-comment-run/raw/develop/doc_assets/saved-replies.gif)

**Comment by nwtgck** [comment]
I think `@github-actions comment` is better. When we want to select what text should be posted or not, we can program in an issue comment for our use, or use my snippet above.

How about`@github-actions comment` to comment only stdout only and `@github-actions comment --stderrr` to comment both stdout and stderr?

**Comment by ibnesayeed** [comment]
I have implemented a special temporary file-based solution #149 which seems to work great cross-language. This avoids a great deal of boilerplate code from every non-JS code. Please feel free to merge the PR or suggest changes.

**Comment by ibnesayeed** [comment]
> I think `@github-actions comment` is better. When we want to select what text should be posted or not, we can program in an issue comment for our use, or use my snippet above.
> 
> How about`@github-actions comment` to comment only stdout only and `@github-actions comment --stderr` to comment both stdout and stderr?

This will be semantically misleading and complicated for users to understand different situations. Besides, it will not allow being selective about only specific messages to be posted. We may not be interested in posting everything from one or the other channel. The file-based solution I implemented in #149 gives great control to users to decide exactly what they want to post from any language (as it is simply a matter of appending/writing to a temporary file), while keeping the signature of runner execution simple as it always has been.

**Comment by nwtgck** [comment]
Thanks. I think both functionalities are important. The users want to comment like #149, and also use GitHub issue comment as stdout. 

The benefit of `@github-actions comment` or `@github-actions run --comment` (syntax is not decided yet) is easiness. The user can output the hello-world printing out to GitHub issue comment. They don't need to open files and close them.

**Comment by ibnesayeed** [comment]
If you do want to support flags like `--comment-stdout` and/or `--comment-stderr` in addition to the buffer file I implemented, that would be cool too. In fact my file-based solution can provide an easy base to implement these flags. When these optional flags are present, we can open the special temporary comment buffer file and *tee* stdout and/or stderr streams (branch out the stream to redirect one branch) to the buffer file. Rest will be taken care of automatically.
